### ISSUE TITLE
ArgumentException when running diagnostic service on .net core

### ISSUE THREAD

**Issue by genlu** [dotnet/roslyn#46687]
Only happens when runing on servicehub core, don't repro on desktop. Deployed code is based on this PR #46026

servicehub log

```
08/10/2020 14:59:06 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:06 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:06 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:07 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:07 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:11 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:17 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120
08/10/2020 14:59:21 Pacific Standard Time: Error : 1 :[dotnet:27160] Unexpected exception: System.ArgumentException: Argument cannot be empty. (Parameter 'analyzers')
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyAnalyzersArgumentForStaticApis(ImmutableArray`1 analyzers, Boolean allowDefaultOrEmpty) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 176
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.VerifyArguments(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 161
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 127
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers..ctor(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\CompilationWithAnalyzers.cs:line 121
   at Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerExtensions.WithAnalyzers(Compilation compilation, ImmutableArray`1 analyzers, CompilationWithAnalyzersOptions analysisOptions) in C:\Users\gel\roslyn2\src\Compilers\Core\Portable\DiagnosticAnalyzer\DiagnosticAnalyzerExtensions.cs:line 34
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<CreateCompilationWithAnalyzersCacheEntryAsync>g__CreateCompilationWithAnalyzerAsync|13_0(Project project, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 262
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.CreateCompilationWithAnalyzersCacheEntryAsync(Project project, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 228
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.<>c__DisplayClass12_0.<<GetOrCreateCompilationWithAnalyzersAsync>g__GetOrCreateCacheEntryAsync|0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 201
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetOrCreateCompilationWithAnalyzersAsync(Project project, Boolean isDocumentAnalysis, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 185
   at Microsoft.CodeAnalysis.Remote.Diagnostics.DiagnosticComputer.GetDiagnosticsAsync(IEnumerable`1 analyzerIds, Boolean reportSuppressedDiagnostics, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\Core\Diagnostics\DiagnosticComputer.cs:line 71
   at Microsoft.CodeAnalysis.Remote.CodeAnalysisService.<>c__DisplayClass17_0.<<CalculateDiagnosticsAsync>b__0>d.MoveNext() in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Services\CodeAnalysisService_Diagnostics.cs:line 41
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Remote.ServiceBase.RunServiceAsync(Func`1 callAsync, CancellationToken cancellationToken) in C:\Users\gel\roslyn2\src\Workspaces\Remote\ServiceHub\Shared\ServiceBase.cs:line 120

```

**Comment by genlu** [comment]
FYI @mavasani

**Comment by mavasani** [comment]
> Only happens when runing on servicehub core, don't repro on desktop

Are you somehow passing in empty analyzers? I can handle a graceful bail out to return zero diagnostics when queries with zero analyzers, but I think there might be something on your side to investigate why there are no analyzers being passed in. Note that the current implementation for all invocations into remote side has validation on in-proc side to not even call into the remote side for no analyzers or all suppressed analyzers case.

**Comment by mavasani** [comment]
More specifically: https://github.com/dotnet/roslyn/blob/6afb1ee022bb68af0aff2df2cadd3069022af7a2/src/Features/Core/Portable/Diagnostics/DiagnosticArguments.cs#L86

**Comment by mavasani** [comment]
Paging in @tmat. I had an offline sync with Gen, and the issue seems to be happening in the `DefaultAnalyzerAssemblyLoader` when executed on core: https://github.com/dotnet/roslyn/blob/6afb1ee022bb68af0aff2df2cadd3069022af7a2/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs#L41 

![image](https://user-images.githubusercontent.com/10605811/89842366-43279b80-db2a-11ea-8211-1da58e2ee81c.png)

Basically, we export the compiler diagnostic analyzer from `Microsoft.CodeAnalysis.CSharp.dll` and mark it as an analyzer asset from Roslyn VSIX: https://github.com/dotnet/roslyn/blob/6afb1ee022bb68af0aff2df2cadd3069022af7a2/src/Compilers/CSharp/Portable/Compilation/CSharpCompilerDiagnosticAnalyzer.cs#L11-L15
https://github.com/dotnet/roslyn/blob/471e7a5131531fece4e59ffed463d258261e24c3/src/VisualStudio/Setup/source.extension.vsixmanifest#L61

Attempting to load analyzer assembly `Microsoft.CodeAnalysis.CSharp.dll` using the `DefaultAnalyzerAssemblyLoader` succeeds on desktop, but seems to fail on core with the assembly already loaded exception @genlu mentioned above. We use `ShadowCopyAnalyzerAssemblyLoader` for NuGet analyzer assemblies and `DefaultAnalyzerAssemblyLoader` for VSIX based ones, which seems reasonable as we don't expect assemblies from VSIX to be changed during a VS session (not sure that is true with OOP mode though).

**Comment by genlu** [comment]
@tmat thoughts?

**Comment by tmat** [comment]
`DefaultAnalyzerAssemblyLoader.Core.cs` uses the load context that Roslyn assemblies are loaded into. When loading an assembly from a path it does not check whether given assembly has already been loaded or not. It makes sense then that if we try to load a Roslyn assembly again it fails with the above error. Seems like we are missing quite a bit of test coverage here.

**Comment by genlu** [comment]
@tmat @sharwell is trying to enable workspace tests on core, I assume that would cover this?
https://github.com/dotnet/roslyn/pull/45813

**Comment by tmat** [comment]
@genlu There are multiple possible fixes depending on what we want our policy of loading analyzer assemblies and their dependencies to be.
One options would be that each analyzer package (directory) has its own AssemblyLoadContext and its dependencies are loaded to this context and thus isolated from other analyzers. Roslyn assemblies would be loaded to a shared context since all analyzers need Roslyn types.

**Comment by mavasani** [comment]
@genlu @sharwell I have tagged this to be discussed at today's meeting so @genlu can be unblocked here. @tmat Would you be able to attend the design meeting today?

**Comment by JoeRobich** [comment]
In this week's design meeting, it was decided this issue didn't need review.

**Comment by mavasani** [comment]
@sharwell @tmat do we have conclusion on this one?

**Comment by mavasani** [comment]
> In this week's design meeting, it was decided this issue didn't need review.

@JoeRobich This was marked for design review based on @tmat's comment here: https://github.com/dotnet/roslyn/issues/46687#issuecomment-674320921. Do you recall why this was marked as not needing review?

**Comment by sharwell** [comment]
The analyzer driver on .NET Core needs to create an Assembly Load Context (ALC) for each unique directory containing one or more analyzer references. This should eliminate the conflict.

**Comment by JoeRobich** [comment]
> Do you recall why this was marked as not needing review?

@mavasani In the meeting @sharwell said that it seemed like an implementation detail and he had a proposal for the compiler team.

> The analyzer driver on .NET Core needs to create an Assembly Load Context (ALC) for each unique directory containing one or more analyzer references. 

Also, recommend using the AssemblyDepenencyResolver within the custom load context. This is a change (https://github.com/dotnet/format/pull/1128) we recently made in dotnet-format to load the Razor source generator dependencies as they no longer ship in the same folder. This class will load them from the "deps.json" that is packaged with the assembly being loaded.

**Comment by genlu** [comment]
Hey @mavasani, any update on this?

**Comment by genlu** [comment]
The proposed ALC change is a duplicate of #52177, and is impemented in #55098
